src/distributions/range.rs

```diff
@@ -47,126 +45,451 @@ use distributions::Distribution;
+/// Helper trait for creating objects using the correct implementation of
+/// `RangeImpl` for the sampling type; this enables `Range::new(a, b)` to work.
+pub trait SampleRange: PartialOrd+Sized {
+    type T: RangeImpl<X = Self>;
 }
 
-macro_rules! integer_impl {
-    ($ty:ty, $unsigned:ident) => {
+/// Helper trait handling actual range sampling.
+///
+/// If you want to implement `Range` sampling for your own type, then
+/// implement both this trait and `SampleRange`:
+///
+/// ```rust
+/// use rand::{Rng, thread_rng};
+/// use rand::distributions::Distribution;
+/// use rand::distributions::range::{Range, SampleRange, RangeImpl, RangeFloat};
+///
+/// #[derive(Clone, Copy, PartialEq, PartialOrd)]
+/// struct MyF32(f32);
+///
+/// #[derive(Clone, Copy, Debug)]
+/// struct RangeMyF32 {
+///     inner: RangeFloat<f32>,
+/// }
+/// impl RangeImpl for RangeMyF32 {
+///     type X = MyF32;
+///     fn new(low: Self::X, high: Self::X) -> Self {
+///         RangeMyF32 {
+///             inner: RangeFloat::<f32>::new(low.0, high.0),
+///         }
+///     }
+///     fn new_inclusive(low: Self::X, high: Self::X) -> Self {
+///         RangeImpl::new(low, high)
+///     }
+///     fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Self::X {
+///         MyF32(self.inner.sample(rng))
+///     }
+/// }
+///
+/// impl SampleRange for MyF32 {
+///     type T = RangeMyF32;
+/// }
+///
+/// let (low, high) = (MyF32(17.0f32), MyF32(22.0f32));
+/// let range = Range::new(low, high);
+/// let x = range.sample(&mut thread_rng());
+/// ```
+pub trait RangeImpl: Sized {
+    /// The type sampled by this implementation.
+    type X: PartialOrd;
+
+    /// Construct self, with inclusive lower bound and exclusive upper bound
+    /// `[low, high)`.
+    ///
+    /// Usually users should not call this directly but instead use
+    /// `Range::new`, which asserts that `low < high` before calling this.
+    fn new(low: Self::X, high: Self::X) -> Self;
+
+    /// Construct self, with inclusive bounds `[low, high]`.
+    ///
+    /// Usually users should not call this directly but instead use
+    /// `Range::new`, which asserts that `low < high` before calling this.
+    fn new_inclusive(low: Self::X, high: Self::X) -> Self;
+
+    /// Sample a value.
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Self::X;
+
+    /// Sample a single value uniformly from a range with inclusive lower bound
+    /// and exclusive upper bound `[low, high)`.
+    /// Panics if `low >= high`.
+    fn sample_single<R: Rng + ?Sized>(low: Self::X, high: Self::X, rng: &mut R)
+        -> Self::X
+    {
+        let range: Self = RangeImpl::new(low, high);
+        range.sample(rng)
+    }
+}
+
+/// Implementation of `RangeImpl` for integer types.
+#[derive(Clone, Copy, Debug)]
+pub struct RangeInt<X> {
+    low: X,
+    range: X,
+    zone: X,
+}
+
+macro_rules! range_int_impl {
+    ($ty:ty, $signed:ident, $unsigned:ident,
+     $i_large:ident, $u_large:ident) => {
```

Comment:
All types should be `ty`, not `ident`.

Reply:
The names are also used like `::core::$u_large::MAX`. `ident` works for both, but `ty` only for types.